src/Voron/Data/Tables/IndexDefinitions.cs

```diff
@@ -0,0 +1,443 @@
+                indexDef.StartIndex = *(int*)currentPtr;
+
+                currentPtr = input.Read(2, out currentSize);
+                indexDef.Count = *(int*)currentPtr;
+
+                currentPtr = input.Read(3, out currentSize);
+                indexDef.IsGlobal = Convert.ToBoolean(*currentPtr);
+
+                currentPtr = input.Read(4, out currentSize);
+                Slice.From(context, currentPtr, currentSize, ByteStringType.Immutable, out indexDef.Name);
+
+                return indexDef;
+            }
+        }
+
+        public class DynamicKeyIndexDef : AbstractTreeIndexDef
+        {
+            public override TreeIndexType Type => TreeIndexType.DynamicKeyValues;
+
+            public delegate ByteStringContext.Scope IndexEntryKeyGenerator(ByteStringContext context, ref TableValueReader value, out Slice slice);
+
+            public IndexEntryKeyGenerator GenerateKey;
+
+            public override ByteStringContext.Scope GetValue(ByteStringContext context, ref TableValueReader value,
+                out Slice slice)
+            {
+                return GenerateKey(context, ref value, out slice);
+            }
+
+            public override ByteStringContext.Scope GetValue(ByteStringContext context, TableValueBuilder value,
+                out Slice slice)
+            {
+                using (context.Allocate(value.Size, out var buffer))
+                {
+                    value.CopyTo(buffer.Ptr);
+                    var reader = value.CreateReader(buffer.Ptr);
+                    return GenerateKey(context, ref reader, out slice);
+                }
+            }
+
+            public override byte[] Serialize()
+            {
+                // We serialize the Type enum as ulong to be "future-proof"
+                var castedType = (long)Type;
+
+                var serializer = new TableValueBuilder
+                {
+                    castedType,
+                    IsGlobal,
+                    Name
+                };
+
+                var methodNameBytes = Encodings.Utf8.GetBytes(GenerateKey.Method.Name);
+                fixed (byte* ptr = methodNameBytes)
+                {
+                    serializer.Add(ptr, methodNameBytes.Length);
+                }
+
+                Debug.Assert(GenerateKey.Method.DeclaringType?.FullName != null && GenerateKey.Method.DeclaringType.Assembly.FullName != null,
+                    $"Invalid {nameof(GenerateKey)} '{GenerateKey.Method.Name}'");
+
+                var assemblyName = new AssemblyName(GenerateKey.Method.DeclaringType.Assembly.FullName);
+                var declaringType = $"{GenerateKey.Method.DeclaringType.FullName}, {assemblyName.Name}";
+                var declaringTypeBytes = Encodings.Utf8.GetBytes(declaringType);
+                fixed (byte* ptr = declaringTypeBytes)
+                {
+                    serializer.Add(ptr, declaringTypeBytes.Length);
+                }
```

Comment:
Why do we care about this at all?

Reply:
Also, does it mean if I change the name of the method we will fail here? but if I will change the body it is okay?

Reply:
We care about that, yes. Because we want to ensure that we aren't messing up the index.
In general, we would only modify that using schema changes, but it allows us to make more complex decisions later (such as indexes that we can select their form at creation time).

And the idea is that you can change the implementation if needed (but likely will require schema change).

Mostly, to be honest, to ensure that we aren't trying to use an index with the wrong delegate.